crates/jit/src/instantiate.rs

```diff
@@ -80,60 +82,103 @@ pub struct CompilationArtifacts {
     debug_info: bool,
 }
 
+// A wrapper to hash compiler flags/tunables and data for cache key.
+struct HashedCompilationArtifactsEnv<'a>(&'a Compiler, &'a [u8]);
+
+impl<'a> Hash for HashedCompilationArtifactsEnv<'a> {
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        // Hash compiler's flags: compilation strategy, isa, frontend config,
+        // misc tunables.
+        let compiler = self.0;
+        compiler.strategy().hash(hasher);
+        compiler.isa().triple().hash(hasher);
+        // TODO: if this `to_string()` is too expensive then we should upstream
+        // a native hashing ability of flags into cranelift itself, but
+        // compilation and/or cache loading is relatively expensive so seems
+        // unlikely.
+        compiler.isa().flags().to_string().hash(hasher);
+        compiler.frontend_config().hash(hasher);
+        compiler.tunables().hash(hasher);
+
+        // TODO: ... and should we hash anything else? There's a lot of stuff in
+        // `TargetIsa`, like registers/encodings/etc. Should we be hashing that
+        // too? It seems like wasmtime doesn't configure it too too much, but
+        // this may become an issue at some point.
+
+        // Hash entire module code + data
+        self.1.hash(hasher);
+    }
+}
+
 impl CompilationArtifacts {
     /// Builds compilation artifacts.
-    pub fn build(compiler: &Compiler, data: &[u8]) -> Result<Self, SetupError> {
-        let environ = ModuleEnvironment::new(compiler.frontend_config(), compiler.tunables());
-
-        let translation = environ
-            .translate(data)
-            .map_err(|error| SetupError::Compile(CompileError::Wasm(error)))?;
-
-        let debug_info = compiler.tunables().debug_info;
-
-        let mut debug_data = None;
-        if debug_info {
-            // TODO Do we want to ignore invalid DWARF data?
-            debug_data = Some(read_debuginfo(&data)?);
-        }
-
-        let Compilation {
-            obj,
-            unwind_info,
-            traps,
-            stack_maps,
-            address_transform,
-        } = compiler.compile(&translation, debug_data)?;
+    pub fn build(
+        compiler: &Compiler,
+        data: &[u8],
+        cache_config: &CacheConfig,
+    ) -> Result<Self, SetupError> {
+        let cache_entry = ModuleCacheEntry::new("wasmtime", cache_config);
+        let result = cache_entry.get_data(
+            HashedCompilationArtifactsEnv(compiler, data),
+            build_artifacts,
+        )?;
+        Ok(result)
+    }
```

Comment:
Would it be possible to do this in the `wasmtime` crate itself? The caching architecture is a pretty high-level concern so I think it'd be best if we could migrate it out of the `wasmtime-environ` crate (or maybe even migrate it to its own crate). It'd be pretty neat if this could be:

```rust
let result = cache_entry.get_data((compiler, wasm), CompilationArtifacts::build);
```

(in wasmtime)

Reply:
done